backend/config/config_test.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+)
+
+func TestNewDefaultAppConfigLanguage(t *testing.T) {
+	tt := []struct{ locale, wantLang string }{
+		{"", "en"},
+		{"-", "en"},
+		{"en", "en"},
+		{"en-US", "en-US"},
+		{"en_US", "en-US"},
+		{"en-something", "en"},
+		{"de", "de"},
+		{"de-DE", "de-DE"},
+		{"de-CH", "de-CH"},
+	}
+	for _, test := range tt {
+		appconf := NewDefaultAppConfig(test.locale)
+		if v := appconf.Frontend["userLanguage"].(string); v != test.wantLang {
+			t.Errorf("%s: userLanguage=%q; want %s", test.locale, v, test.wantLang)
+		}
```

Comment:
since we are using testify everywhere else, you could save some lines and manual error msg formatting by using it, e.g. `require.Equal(t, test.wantLang, appconf.Frontend["userLanguage"].(string))`.

Reply:
Here's my partially imagined experience with testify today. Took some notes:

```go
tt := []struct{ locale, wantLang string }{
  {"de DE", "de-DE1"},
  {"de-CH", "de-CH2"},
}
```

(wantLang is wrong to make tests fail on purpose)

- 10 extra lines of useless text per test case:

```
$ go test # with testify
--- FAIL: TestNewDefaultAppConfigLanguage (0.00s)
    config_test.go:39:
                Error Trace:    config_test.go:39
                Error:          Not equal:
                                expected: "en"
                                actual  : "de-DE1"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -en
                                +de-DE1
                Test:           TestNewDefaultAppConfigLanguage
```

That's exactly 100 extra lines of useless text in the worst case to dig through if I have 10 cases. ...yeah, I guess it can always get worse. Let's keep an open mind.
 
Although, all I really need, with regular testing pkg, is this:

```
--- FAIL: TestNewDefaultAppConfigLanguage (0.00s)
    config_test.go:39: de DE: userLanguage="en"; want de-DE1
    config_test.go:39: de-CH: userLanguage="de-CH"; want de-CH2
```

With testify, I see only values, no context. In larger tests it plays an even more important role.

Ok, I guess I can add another arg to require.Equal. Let's see:

```go
require.Equal(t, appconf.Frontend["userLanguage"].(string), test.wantLang, "userLanguage")
```

Apart from the line now being longer and I might not see it all on the screen so maybe opt for split on 2 lines instead, here's what I see:

```
--- FAIL: TestNewDefaultAppConfigLanguage (0.00s)
    config_test.go:39:
                Error Trace:    config_test.go:39
                Error:          Not equal:
                                expected: "de-DE1"
                                actual  : "en"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -de-DE1
                                +en
                Test:           TestNewDefaultAppConfigLanguage
                Messages:       userLanguage
```

A message! What does it mean? Who knows. I'll go read the test code instead of fixing the problem...

Maybe I'll end up writing a longer message like "if you see this, it means userLanguage output is incorrect" so next time someone sees it, they'll know what it means!

Hold on, expected vs actual. Which is which? go doc require Equal:

```go
func Equal(t TestingT, expected interface{}, actual ...
```

Aha! Need to swap the args. Ok, for the damned 10th time, I really hope I'll remember. It's the opposite of the convention which I see every day when reading normal testing pkg tests and always get confused. I hope Python people are not reading this.

Anyway, notice how regular tests above displayed two errors, not one, which brings me to:

- `require.Equal` is actually `t.Fatal` but I need `t.Error`.

Seeing all errored test cases often tells a pattern of what's wrong, unlike only the first failure and you start wondering, "but if the input were _this_, what would've been the output?". That's why I don't normally roll the dice between `t.Fatal` and `t.Error` and pick the one that fits best.

I guess I'll go and temporary comment out some lines to let the test keep going instead of having possibly already understood the problem...

For example, in TDD you start from nothing, add test cases, keep it running and implement until it's all green. Repeat. Basically imposible with testify because it fails on first error.

- Ok, maybe there's a `require.EqualButNotFatal`. Let's see:

```
$ go doc require | grep -E ^func | wc -l
105
```

That's how many require.Xxx funcs there are. I guess I'll go read them instead of writing a simple test...

Ok, I stopped reading. That's too many.

- Sure, I can always t.Run. This also improves on t.Fatal and the TDD problem:

```go
t.Run(test.locale, func(t *testing.T) {
  // test code here
})
```

```
--- FAIL: TestNewDefaultAppConfigLanguage (0.00s)
    --- FAIL: TestNewDefaultAppConfigLanguage/de_DE (0.00s)
        config_test.go:40:
                Error Trace:    config_test.go:40
                Error:          Not equal:
                                expected: "de-DE1"
                                actual  : "en"
```

Wait a second, actual "en" from input "de_DE"? That's not good at all.
Oh man, of course, did I forget to shadow the loop var? Maybe.

```go
test := test
```

... nope. Same result.

... after 1 hour:

Totally forgot that `t.Run(name, func)` replaces spaces with underscores in the name! So, the input was actually "de DE", with a space. Makes sense now why the actual output is "en".

Ok, I guess I'll have to add the input to my already long message `... output is incorrect; the input was %q", test.locale`.

So, what's `t.Run`'s test name? A confusing artifact of me being forced to use subtests. Otherwise, it gets even worse due to t.Fatal everywhere.

Reply:
If you really insist, I'll replace with testify but I hope the above justifies why I resist.

Reply:
I didn't realize even https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures mentions it.

Reply:
> Ok, maybe there's a `require.EqualButNotFatal.` Let's see: ... Repeat. Basically imposible with testify because it fails on first error.

You can use `assert.Equal` instead of `require.Equal` if you don't wish to abort.

> A message! What does it mean? Who knows. I'll go read the test code instead of fixing the problem...

That's normal? If a test fails, you open the test code first in any case. Never had an issue identifying the problem with testify.

> Wait a second, actual "en" from input "de_DE"? That's not good at all.

Sorry I don't think I follow what's happening here anymore ;) Explain over :beer: 

In any case, of course I don't resist ;) but I also think you dismiss it too early. Longer error output in the test can be good and bad, but mostly good. More concise test code is good imho. Let's convert this for fun together one day and discuss pros and cons.

Maybe something other than testify also does the job in a way you like better, but adding at least three lines of code and manual formatting of error mesages everytime I want to write one assertion is almost a no-go for me.

Reply:
I guess the main difference is I prefer to not even read the test code and rely on the easy-to-read test output so I can go directly to the actual source code and fix the problem. Testify and other assert-like packages get very much in the way of that.

Source code: must be readable
Test code: don't care as much but its **output** must be easy readable 

You rely on the test source code and not so much on the test output. As in, for you the test output is more like a hint.